streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -185,12 +186,31 @@ private void cleanup(final long segmentId) {
         }
     }
 
-    private long segmentIdFromSegmentName(String segmentName) {
-        try {
-            Date date = formatter.parse(segmentName.substring(name.length() + 1));
-            return date.getTime() / segmentInterval;
-        } catch (Exception ex) {
-            return -1L;
+    private long segmentIdFromSegmentName(String segmentName, final File parent) {
+        // old style segment name with date
+        if (segmentName.charAt(name.length()) == '-') {
+            final String datePart = segmentName.substring(name.length() + 1);
+            final Date date;
+            try {
+                date = formatter.parse(datePart);
+            } catch (ParseException e) {
+                throw new ProcessorStateException("Unable to parse segmentName:"
```

Comment:
Before the change, we did return `-1L` but did not throw -- How does throwing an exception influence the exception path ?

Reply:
I was wondering why we returned `-1L`. I don't think it is valid, i.e, what are you actually opening? It could be an invalid directory in which case there would be an exception when the database opened. If it isn't invalid and there are multiple directories that fail with this then all the segments are `-1L` which isn't correct either. So... I thought it would be better to bail here as the state directory must have some invalid or tinkered with dirs in it

Reply:
I agree that returning `-1L` does not seems to be correct. I am just wondering, if we handle this new exception properly as we change error handling with this change. Just try to understand the impact and if we need to do some other changes somewhere else or not.

Reply:
Originally if parse exception happens we simply do not load this segment at all, which is simply a bad design as a hindsight. This change is much better and we just need to remember handle the thrown exception in some higher call level.

However, I'm wondering if the folder would ever contain any other files besides the segment database folders like `.LOCK` files, etc? If that could be true, maybe we should log-and-skip than throwing or silently-skip.

Reply:
I'm not sure it would contain other files, but i guess it could. Maybe logging a warning and moving on is ok

Reply:
> I'm not sure it would contain other files, but i guess it could. Maybe logging a warning and moving on is ok

Cool